File: Shashank Shukla 18-CSE/Day-6.java
/*
Time Complexity: O(N)
Space Complexity: O(N)
Approach: Using single stack store the character and it's count as a pair in a stack and as count becomes equal to k remove the character.
*/

class Pair{
    char val;
    int count;
    public Pair(char c, int v){
        this.val = c;
        this.count = v;
    }
}

class Solution {
    public String removeDuplicates(String s, int k) {
        Stack<Pair> st = new Stack<>();
        
        for(int i=0; i<s.length(); i++){
            if(st.isEmpty())
                st.push(new Pair(s.charAt(i), 1));
            else{
                Pair cur = st.peek();
                if(cur.val != s.charAt(i)){
                    st.push(new Pair(s.charAt(i), 1));
                } else {
                    int ctr = cur.count;
                    if(ctr+1 == k)
                        st.pop();
                    else 
                        cur.count = ctr + 1;
                }
            }
            
        }
        String res="";
        while(!st.isEmpty()){
            Pair ans = st.pop();
            while(ans.count-- > 0)
                res = ans.val + res;
        }
        return res;
    }
}

File: Aman Dwivedi 18-ECE/Day 6.cpp
https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string-ii/
Time: O(N)
Space: O(N)

class Solution {
public:
    string removeDuplicates(string s, int k) {

        int n=s.size();
        stack<pair<char,int>> st; //take frequency stack to store every element
        for(int i=0;i<n;i++)
        {
            //if first char is not there or equal to top, then add 
            if(st.empty()||st.top().first!=s[i])
                st.push({s[i],1});
            else
            {
                //if it is there we increase its count
                auto ele=st.top();
                st.pop();
                st.push({s[i],ele.second+1});
            }
            //if its freq == k, then we remove it
            if(st.top().second==k)
                st.pop();
        }
        //take empty string to store result
        string res="";
        while(!st.empty())
        {
            while(st.top().second--)
                res+=st.top().first;

            st.pop();
        }
        //as chars are stored in reverse order, we reverse the string
        reverse(res.begin(),res.end());
        return res;

    }
};g
